lib/PhysicalobjectCsvImporter.class.php

```diff
@@ -0,0 +1,655 @@
+   * @return void
+   */
+  public function createKeymapEntry($object, $csvdata)
+  {
+    $keymap = new $this->ormClasses['keymap'];
+    $keymap->sourceName = $this->getOption('sourceName');
+
+    if (!empty($csvdata['legacyId']))
+    {
+      $keymap->sourceId = $csvdata['legacyId'];
+    }
+
+    // Determine target name using object class
+    $keymap->targetName = sfInflector::underscore(
+      str_replace('Qubit', '', get_class($object)));
+    $keymap->targetId   = $object->id;
+
+    $keymap->save();
+  }
+
+  protected function log($msg)
+  {
+    // Just echo to STDOUT for now
+    echo $msg.PHP_EOL;
+  }
+
+  protected function logError($msg)
+  {
+    fwrite($this->getErrorLogHandle(), $msg.PHP_EOL);
+  }
+
+  protected function progressUpdate($data)
+  {
+    $freq = $this->getProgressFrequency();
+
+    if (1 == $freq)
+    {
+      $this->log(sprintf('Imported row [%u/%u]: name "%s"',
+        $this->offset, $this->rowsTotal, $data['name']));
+    }
+    else if ($freq > 1 && 0 == $this->rowsImported % $freq)
+    {
+      $this->log(sprintf('Imported %u of %u rows...',
+        $this->rowsImported, $this->rowsTotal));
+    }
+  }
+
+  protected function getDbConnection()
+  {
+    if (null === $this->dbcon)
+    {
+      $this->dbcon = Propel::getConnection();
+    }
+
+    return $this->dbcon;
+  }
+
+  protected function getErrorLogHandle()
+  {
+    if (null === $filename = $this->getOption('errorLog'))
+    {
+      return STDERR;
+    }
+
+    if (!isset($this->errorLogHandle))
+    {
+      $this->errorLogHandle = fopen($filename, 'w');
+    }
+
+    return $this->errorLogHandle;
+  }
+
+  protected function readCsvFile($filename)
+  {
+    $reader = \League\Csv\Reader::createFromPath($filename, 'r');
+
+    if (!isset($this->options['header']))
+    {
+      // Use first row of CSV file as header
+      $reader->setHeaderOffset(0);
+    }
+
+    $this->rowsTotal = count($reader);
+
+    return $reader;
+  }
+
+  protected function getRecords($stmt)
+  {
+    if (isset($this->options['header']))
+    {
+      $records = $stmt->process($this->reader, $this->options['header']);
+    }
+    else
+    {
+      $records = $stmt->process($this->reader);
+    }
+
+    return $records;
+  }
+
+  protected function processColumn($key, $val, $culture)
+  {
+    switch ($key)
+    {
+      case 'culture':
+        $val = $culture;
+
+        break;
+
+      case 'type':
+        $val = $this->lookupTypeId($val, $culture);
+
+        break;
+
+      case 'descriptionSlugs':
+        $val = $this->processDescriptionSlugs($val);
+
+        break;
+
+      default:
+        $val = trim($val);
+    }
+
+    if ('' !== $val)
+    {
+      return $val;
+    }
+  }
+
+  protected function processDescriptionSlugs(string $str = null)
+  {
+    $ids = [];
+
+    if (null === $str)
+    {
+      return $ids;
+    }
+
+    foreach ($this->processMultiValueColumn($str) as $val)
+    {
+      $class = $this->ormClasses['informationObject'];
```

Comment:
I don't understand the benefit of this technique (adding the classes to an array instead of using the class name directly).  In this case it would turn 2 lines of code into one.

Reply:
Yes, I'm using the ormClasses array to [inject mock ORM objects](https://github.com/artefactual/atom/blob/e973cfd323beb83b1bed8f3991f258ef89b60721/test/phpunit/PhysicalobjectCsvImporterTest.php#L24) when running tests to avoid connecting to the database and avoid side effects from the stated of the database data.

Reply:
Ah! I've love a tour of all this and how it fits together - so many interesting changes!

Reply:
@sbreker I'm happy to do a tour.  I'm taking care of my kids today (Pro-D day) but maybe on Monday I can take you through it. :)